mirascope/gemini/calls.py

```diff
@@ -0,0 +1,167 @@
+"""A module for calling Google's Gemini Chat API."""
+import datetime
+import logging
+from typing import (
+    Any,
+    AsyncGenerator,
+    ClassVar,
+    Generator,
+)
+
+from google.generativeai import GenerativeModel  # type: ignore
+from google.generativeai.types import ContentsType  # type: ignore
+
+from ..base import BaseCall
+from .tools import GeminiTool
+from .types import (
+    GeminiCallParams,
+    GeminiCallResponse,
+    GeminiCallResponseChunk,
+)
+
+logger = logging.getLogger("mirascope")
+
+
+class GeminiCall(BaseCall[GeminiCallResponse, GeminiCallResponseChunk, GeminiTool]):
+    '''A class for prompting Google's Gemini Chat API.
+
+    This prompt supports the message types: USER, MODEL, TOOL
+
+    Example:
+
+    ```python
+    from google.generativeai import configure  # type: ignore
+    from mirascope.gemini import GeminiPrompt
+
+    configure(api_key="YOUR_API_KEY")
+
+
+    class BookRecommender(GeminiPrompt):
+        prompt_template = """
+        USER: You're the world's greatest librarian.
+        MODEL: Ok, I understand I'm the world's greatest librarian. How can I help?
+        USER: Please recommend some {genre} books.
+
+        genre: str
+
+
+    response = BookRecommender(genre="fantasy").call()
+    print(response.call())
+    #> As the world's greatest librarian, I am delighted to recommend...
+    ```
+    '''
+
+    call_params: ClassVar[GeminiCallParams] = GeminiCallParams(
+        model="gemini-1.0-pro",
+        generation_config={"candidate_count": 1},
+    )
+
+    @property
+    def messages(self) -> ContentsType:
+        """Returns the `ContentsType` messages for Gemini `generate_content`.
+
+        Raises:
+            ValueError: if the docstring contains an unknown role.
+        """
+        return [
+            {"role": message["role"], "parts": [message["content"]]}
+            for message in self._parse_messages(["model", "user", "tool"])
+        ]
+
+    def call(self, **kwargs: Any) -> GeminiCallResponse:
+        """Makes an call to the model using this `GeminiCall` instance.
+
+        Args:
+            **kwargs: Additional keyword arguments that will be used for generating the
+                response. These will override any existing argument settings in call
+                params.
+
+        Returns:
+            A `GeminiCallResponse` instance.
+        """
+        kwargs, tool_types = self._setup(kwargs, GeminiTool)
+        gemini_pro_model = GenerativeModel(kwargs.pop("model"))
```

Comment:
GenerativeModel takes in `model_name` instead of `model`. Our call params for GeminiCallParams uses `model`. Are we basically saying that we will always use `gemini-pro` and that `gemini-1.0-pro` which is our default isn't being used?

Reply:
it's not keyworded, so it'll set model_name to `kwargs.pop("model")` which will come either from the kwargs the user provides OR the call params (kwargs take priority as overrides)

Reply:
I'm updating it to keyword `model_name=kwargs.pop("model_name")` for consistency and updating the gemini call params to use `model_name` instead of `model`

Reply:
Scratch that lol can't use `model_name` it's protected

Reply:
I'm still setting it as a keyword arg now tho for clarity

Reply:
gotcha